{
  "className":"sprouts.Result_Spec",
  "title":"Results",
  "narrative":"The `Result` interface is used to represent the optional result of an operation\n    as well as a list of problems that occurred during the operation.\n\n    It is a sub type of the `Val` property and as such can be used to represent a value\n    that is immutable and can be observed for changes.\n\n    The default result implementation indirectly exposed by its factory methods\n    is immutable and thread safe, effectively making it a monadic value\n    similar to the `Optional` type in Java 8.",
  "subjects":["sprouts.Result","sprouts.Val","sprouts.Problem"],
  "statistics":{
    "runs":"15",
    "successRate":"100.0%",
    "failures":"0",
    "errors":"0",
    "skipped":"0",
    "duration":"?"
  },
  "headers":[],"tags":{},"see":[],
  "features":[ 
    {
      "id":"We can create a result from any kind of value.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":[]
      },
      "blocks":[
        {"kind":"expect","text":"","code":["Result.of(42).get() == 42","Result.of(\"foo\").get() == \"foo\"","Result.of([1, 2, 3]).get() == [1, 2, 3]"]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"A result can be created from multiple problems.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":[]
      },
      "blocks":[
        {"kind":"given","text":"A list of problems.","code":["def problems = [","            Problem.of(\"too large\"),","            Problem.of(\"too even\"),","            Problem.of(\"not a prime\")","        ]"]},

        {"kind":"when","text":"We create a result from the problems.","code":["def result = Result.of(Integer, problems)"]},

        {"kind":"then","text":"The result has the problems.","code":["result.problems() == problems"]},

        {"kind":"and","text":"Although they are the same problems, they are not the same instances.","code":["result.problems() !== problems"]},

        {"kind":"when","text":"We try to mutate the problems.","code":["result.problems().add(null)"]},

        {"kind":"then","text":"An exception is thrown.","code":["thrown(UnsupportedOperationException)"]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"The items of a `Result` can be mapped using mapping functions.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":[]
      },
      "blocks":[
        {"kind":"given","text":"A result.","code":["def result = Result.of(42)"]},

        {"kind":"when","text":"We map the result to another result with the value plus one.","code":["def mapped = result.map { it + 1 }"]},

        {"kind":"then","text":"The mapped result has the value of the result plus one.","code":["mapped.get() == 43"]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"Results can be turned into an Optional.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":[]
      },
      "blocks":[
        {"kind":"given","text":"A result.","code":["def result = Result.of(42)"]},

        {"kind":"when","text":"We turn the result into an Optional.","code":["def optional = result.toOptional()"]},

        {"kind":"then","text":"The optional has the value of the result.","code":["optional.get() == 42"]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"Just like a `Val` property, a `Result` has a type and id.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":[]
      },
      "blocks":[
        {"kind":"given","text":"A result.","code":["def result = Result.of(42)"]},

        {"kind":"expect","text":"The result has a type.","code":["result.type() == Integer"]},

        {"kind":"and","text":"The result has an empty id.","code":["result.id().isEmpty()"]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"You can create a `Result` from a list.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":["\n            A result can be created from a list of values which\n            is useful when an operation has multiple result values.\n        "]
      },
      "blocks":[
        {"kind":"given","text":"A list of values.","code":["def values = [1, 2, 3]"]},

        {"kind":"when","text":"We create a result from the list.","code":["def result = Result.ofList(Integer, values)"]},

        {"kind":"then","text":"The result has the list as its value.","code":["result.get() == values"]},

        {"kind":"and","text":"The result has no problems.","code":["result.problems().isEmpty()"]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"Create a Result from a list with some problems.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":["\n            A result can be created from a list of values which\n            is useful when an operation has multiple result values.\n        "]
      },
      "blocks":[
        {"kind":"given","text":"A list of values.","code":["def values = [1, 2, 3]"]},

        {"kind":"and","text":"A list of problems.","code":["def problems = [","            Problem.of(\"too large\"),","            Problem.of(\"too even\"),","            Problem.of(\"not a prime\")","        ]"]},

        {"kind":"when","text":"We create a result from the list.","code":["def result = Result.ofList(Integer, values, problems)"]},

        {"kind":"then","text":"The result has the list as its value.","code":["result.get() == values"]},

        {"kind":"and","text":"The result has the problems.","code":["result.problems() == problems"]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"An empty `Result` can be mapped to any property type without an exception being thrown.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":["\n            An empty `Result` is similar to an empty `Optional` in that\n            it does not have a value. However, unlike an empty `Optional`,\n            a `Result` instance also has a type.\n            So when you want to map it to some other type, using `mapTo`, you need to\n            also provide the target type.\n        "]
      },
      "blocks":[
        {"kind":"given","text":"An empty result.","code":["def result = Result.of(Integer, null)"]},

        {"kind":"when","text":"We map the result to a property.","code":["def mapped = result.mapTo(String, it -> \"foo $it\" )"]},

        {"kind":"then","text":"The resulting property is also empty.","code":["mapped.isEmpty()"]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"Exceptions inside of a mapping function are caught and turned into problems.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":["\n            Mapping `Result` objects is based on a mapping function\n            which has the responsibility of mapping the item of the result\n            to another item.\n\n            However, sometimes the mapping function might throw an exception\n            in which case the exception is caught and turned into a problem and\n            the resulting property will be an empty result with the problem.\n\n            The reason for this is because the inherent purpose of a `Result` is to\n            protect against exceptions crippling the application control flow\n            while also preserving a record of what went wrong.\n        "]
      },
      "blocks":[
        {"kind":"given","text":"A result.","code":["def result = Result.of(42)"]},

        {"kind":"when","text":"We map the result unsuccessfully by throwing an exception:","code":["def mapped = result.map { throw new IllegalArgumentException(\"foo\") }"]},

        {"kind":"then","text":"The resulting property has a problem.","code":["mapped.problems().size() == 1"]},

        {"kind":"and","text":"The problem is an exception problem.","code":["mapped.problems().first().exception().get().message == \"foo\""]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"If mapping to another type goes wrong, the exception is caught and turned into a problem.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":["\n            Mapping `Result` objects is based on a mapping function\n            which has the responsibility of mapping the item of the result\n            to another item.\n\n            However, sometimes the mapping function might throw an exception\n            in which case the exception is caught and turned into a problem and\n            the resulting property will be an empty result with the problem.\n\n            The reason for this is because the inherent purpose of a `Result` is to\n            protect against exceptions crippling the application control flow\n            while also preserving a record of what went wrong.\n        "]
      },
      "blocks":[
        {"kind":"given","text":"A simple result with a simple item.","code":["def result = Result.of(42)"]},

        {"kind":"when","text":"We map the result to another type with an exception.","code":["def mapped = result.mapTo(String, { throw new IllegalArgumentException(\"foo\") })"]},

        {"kind":"then","text":"The resulting property has a problem.","code":["mapped.problems().size() == 1"]},

        {"kind":"and","text":"The problem is an exception problem.","code":["mapped.problems().first().exception().get().message == \"foo\""]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"Create a result from a supplier which may or may not throw an exception using the `ofTry` method.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":["\n            The core purpose of a `Result` is to resolve a value \n            while being protected against exceptions\n            that might occur during the execution of an operation\n            that produces the value.\n\n            The `ofTry` method is used to create a `Result` from a value supplier.\n            The supplier may or may not throw an exception.\n            If it does, the exception is caught and turned into a problem.\n        "]
      },
      "blocks":[
        {"kind":"given","text":"A flag that determines if an exception should be thrown.","code":["def doFail = false"]},

        {"kind":"and","text":"A supplier that may or may not throw an exception.","code":["Supplier<String> supplier = ()->{","    if ( doFail )","        throw new IllegalArgumentException(\"foo\")","    else","        return \"bar\"","}"]},

        {"kind":"when","text":"We create a result from the supplier with the `doFail` flag set to false.","code":["def result = Result.ofTry(String.class, supplier)"]},

        {"kind":"then","text":"No exception is thrown, which means the result has the value.","code":["result.get() == \"bar\""]},

        {"kind":"and","text":"The result has no problems.","code":["result.problems().isEmpty()"]},

        {"kind":"when","text":"We create a result from the supplier with the `doFail` flag set to true.","code":["doFail = true","def result2 = Result.ofTry(String.class, supplier)"]},

        {"kind":"then","text":"An exception is thrown, which means the result has a problem.","code":["result2.problems().size() == 1"]},

        {"kind":"and","text":"The problem is an exception problem.","code":["result2.problems().first().exception().get().message == \"foo\""]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"You can recognize a `Result` by its String representation.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":["\n            A `Result` instance has a specific string representation that \n            tells you both the type of the result and the current item of the result.\n            The string representation of starts with \"Result\" followed by the type of the result\n            and the item of the result.\n        "]
      },
      "blocks":[
        {"kind":"given","text":"A result object holding a common enum value.","code":["def result = Result.of(DayOfWeek.MONDAY)"]},

        {"kind":"expect","text":"The string representation of the result.","code":["result.toString() == \"Result<DayOfWeek>[MONDAY]\""]},

        {"kind":"when","text":"We update the result with a String id.","code":["result = result.withId(\"foo\")"]},

        {"kind":"then","text":"The string representation of the result reflects the id.","code":["result.toString() == \"Result<DayOfWeek>[foo=MONDAY]\""]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"The equality of two `Result` instances is based on the type and the item of the result.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":["\n            The `Result` type is a value centric type which means that\n            the equality of two `Result` instances is based on the item of the result\n            and its type.\n        "]
      },
      "blocks":[
        {"kind":"expect","text":"The following results are equal and they have the same hash code.","code":["Result.of(42) == Result.of(42)","Result.of(42).hashCode() == Result.of(42).hashCode()","Result.of(Integer) == Result.of(Integer)","Result.of(Integer).hashCode() == Result.of(Integer).hashCode()"]},

        {"kind":"and","text":"The following results are not equal and they have different hash codes.","code":["Result.of(42) != Result.of(43).withId(\"foo\")","Result.of(42).hashCode() != Result.of(43).withId(\"foo\").hashCode()","Result.of(Integer) != Result.of(String)","Result.of(Integer).hashCode() != Result.of(String).hashCode()","Result.of(Integer) != Result.of(Integer).withId(\"foo\")","Result.of(Integer).hashCode() != Result.of(Integer).withId(\"foo\").hashCode()","Result.of(42) != Result.of(Integer)","Result.of(42).hashCode() != Result.of(Integer).hashCode()"]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"A `Result` tells us that it is immutable.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":["\n            A `Result` is an immutable value type which means that\n            once created, it cannot be changed.\n        "]
      },
      "blocks":[
        {"kind":"given","text":"A result.","code":["def result = Result.of(42)"]},

        {"kind":"expect","text":"The `isMutable()` method returns false.","code":["!result.isMutable()"]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    },
  
    {
      "id":"A result will find the correct type of an item, even if it is an anonymous class based enum constant.",
      "result":"PASS",
      "duration":"?",
      "iterations":{
      "tags":{},"see":[],"extraInfo":["\n\n            An interesting little quirk of the Java language is that\n            you can have enum constants of the same enum type but with\n            different `Class` instances!\n            An example of this would be:\n\n            ```java\n                public enum Food {\n                    TOFU { @Override public String toString() { return \"Tofu\"; } },\n                    TEMPEH { @Override public String toString() { return \"Tempeh\"; } },\n                    SEITAN { @Override public String toString() { return \"Seitan\"; } },\n                    NATTO { @Override public String toString() { return \"Natto\"; } }\n                }\n            ```\n            Believe it or not but expressions like `Food.TOFU.getClass() == Food.SEITAN.getClass()`\n            or even `Food.TOFU.getClass() == Food.class` are actually both `false`!\n            This is because the enum constants defined above are actually based\n            on anonymous classes. More specifically this is due to the curly brackets\n            followed after the constants declaration itself.\n\n            This could potentially lead to bugs when creating a result property from such an enum constant.\n            More specifically `Result.of(Food.NATTO).type() == Result.of(Food.class, null)` would lead to \n            being evaluated as false **despite the fact that they both have the same generic type**.\n\n            Don't worry however, Sprouts knows this, and it will account for these kinds of enums.\n        "]
      },
      "blocks":[
        {"kind":"given","text":"We create various `Result` instances using the `Food` enum as a basis.","code":["var nonNull = Result.of(Food.NATTO)","var nullable = Result.of(Food, Food.TEMPEH)"]},

        {"kind":"expect","text":"The type of the result is correctly identified.","code":["nonNull.type() == Food","nullable.type() == Food"]}
      ],
      "problems":{"dataValues":[], "errors":[]}
    }
  
  ],
  "generator":"https://github.com/renatoathaydes/spock-reports"
}
